<h1  style="text-align: center"  class="list-group-item list-group-item-warning"> <span style="color:#2E6AA9">Data Cleaning of Global Terrorism Data</span>  </h1> 

## Introduction

### Overview of the Global Terrorism Database (GTD)

The Global Terrorism Database™ (GTD) is a comprehensive open-source database containing detailed information on terrorist events worldwide, spanning from 1970 through 2020. The database is continuously updated, with additional annual updates planned for the future. Unlike many other event databases, the GTD provides systematic data not only on international terrorist incidents but also on domestic and transnational events. It encompasses a vast collection of over 200,000 cases, each offering valuable insights into acts of terrorism.

The GTD entries include crucial details such as the date and location of the incident, weapons used, the nature of the target, casualties, and, when ascertainable, the responsible group or individual. The statistical information within the GTD is sourced from credible open media reports, ensuring the reliability and accuracy of the data. Users are cautioned not to infer legal actions or outcomes from the information presented in the GTD entries.

The National Consortium for the Study of Terrorism and Responses to Terrorism (START) is committed to providing access to the GTD through this online interface. By doing so, they aim to enhance understanding of terrorist violence, enabling better study and strategies to combat terrorism effectively.

### Characteristics of the GTD

- Contains data on over 200,000 terrorist attacks.
- Represents the most comprehensive unclassified database on terrorist attacks globally.
- Encompasses information on more than 88,000 bombings, 19,000 assassinations, and 11,000 kidnappings since 1970.
- Provides details on at least 45 variables for each case, with recent incidents offering information on more than 120 variables.
- Utilizes a vast corpus of over 4,000,000 news articles and 25,000 news sources to collect incident data from 1998 to 2017 alone.

For more information and to request access to the dataset, visit the [Global Terrorism Database (GTD) page](https://www.start.umd.edu/gtd/about/).

## Let's begin the exploration!

In [1]:
## Importing packages for analysis

import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)   # display all columns without any truncation
pd.set_option('display.expand_frame_repr', False) # the DataFrame will not be displayed in a single line 
                                                    # and may span multiple lines for better readability.
    
pd.set_option('max_colwidth', None)  # the entire content of each column will be displayed without truncation

import matplotlib.pyplot as plt   # for visualisations
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')  # a filter to ignore all warnings generated in the code 

'''ensures that all plots generated 
using Matplotlib will be displayed within the notebook interface as static images '''

%matplotlib inline 

In [2]:
df = pd.read_csv('gtd.csv', encoding = "utf-8") # as the dataset requires a particular encoding
print(f'Our data has {df.shape[0]} rows and {df.shape[1]} columns.')
df.head()

Our data has 209706 rows and 135 columns.


eventid  iyear  imonth  iday approxdate  extended resolution  country         country_txt  region                   region_txt provstate           city   latitude   longitude  specificity  vicinity location summary  crit1  crit2  crit3  doubtterr  alternative alternative_txt  multiple  success  suicide  attacktype1                 attacktype1_txt  attacktype2 attacktype2_txt  attacktype3 attacktype3_txt  targtype1                targtype1_txt  targsubtype1                                      targsubtype1_txt                        corp1                  target1  natlty1         natlty1_txt  targtype2 targtype2_txt  targsubtype2 targsubtype2_txt corp2 target2  natlty2 natlty2_txt  targtype3 targtype3_txt  targsubtype3 targsubtype3_txt corp3 target3  natlty3 natlty3_txt                               gname gsubname gname2 gsubname2 gname3 gsubname3 motive  guncertain1  guncertain2  guncertain3  individual  nperps  nperpcap  claimed  claimmode claimmode_txt  claim2  claimmode2 claimmode2_txt  claim3  claimmode3 claimmode3_txt  compclaim  weaptype1 weaptype1_txt  weapsubtype1        weapsubtype1_txt  weaptype2 weaptype2_txt  weapsubtype2 weapsubtype2_txt  weaptype3 weaptype3_txt  weapsubtype3 weapsubtype3_txt  weaptype4 weaptype4_txt  weapsubtype4 weapsubtype4_txt  weapdetail  nkill  nkillus  nkillter  nwound  nwoundus  nwoundte  property  propextent propextent_txt  propvalue propcomment  ishostkid  nhostkid  nhostkidus  nhours  ndays divert kidhijcountry  ransom  ransomamt  ransomamtus  ransompaid  ransompaidus ransomnote  hostkidoutcome hostkidoutcome_txt  nreleased addnotes scite1 scite2 scite3 dbsource  INT_LOG  INT_IDEO  INT_MISC  INT_ANY related
0  197000000001   1970       7     2        NaN         0        NaN       58  Dominican Republic       2  Central America & Caribbean  National  Santo Domingo  18.456792  -69.951164          1.0         0      NaN     NaN      1      1      1          0          NaN             NaN       0.0        1        0            1                   Assassination          NaN             NaN          NaN             NaN         14  Private Citizens & Property          68.0                                        Named Civilian                          NaN             Julio Guzman     58.0  Dominican Republic        NaN           NaN           NaN              NaN   NaN     NaN      NaN         NaN        NaN           NaN           NaN              NaN   NaN     NaN      NaN         NaN                              MANO-D      NaN    NaN       NaN    NaN       NaN    NaN          0.0          NaN          NaN           0     NaN       NaN      NaN        NaN           NaN     NaN         NaN            NaN     NaN         NaN            NaN        NaN         13       Unknown           NaN                     NaN        NaN           NaN           NaN              NaN        NaN           NaN           NaN              NaN        NaN           NaN           NaN              NaN         NaN    1.0      NaN       NaN     0.0       NaN       NaN         0         NaN            NaN        NaN         NaN        0.0       NaN         NaN     NaN    NaN    NaN           NaN     0.0        NaN          NaN         NaN           NaN        NaN             NaN                NaN        NaN      NaN    NaN    NaN    NaN     PGIS        0         0         0        0     NaN
1  197000000002   1970       0     0        NaN         0        NaN      130              Mexico       1                North America   Federal    Mexico city  19.371887  -99.086624          1.0         0      NaN     NaN      1      1      1          0          NaN             NaN       0.0        1        0            6     Hostage Taking (Kidnapping)          NaN             NaN          NaN             NaN          7      Government (Diplomatic)          45.0  Diplomatic Personnel (outside of embassy, consulate)  Belgian Ambassador Daughter  Nadine Chaval, daughter     21.0             Belgium        NaN           NaN          

Now let us dive deeper into our data and understand it better.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 209706 entries, 0 to 209705
Columns: 135 entries, eventid to related
dtypes: float64(54), int64(23), object(58)
memory usage: 216.0+ MB


- We are unable to get the datatype as well as null value count information using .info() method.
- Let us look for the duplicated values (if present) in our dataset.

In [4]:
if df.duplicated().sum() == 0:
    print('No Duplicate values in the data.')
else:
    print(f'There are {df.duplicated().sum()} values in our data.')

No Duplicate values in the data.


**How does our data look mathematically?**

In [5]:
df.describe()

eventid          iyear         imonth           iday       extended        country         region       latitude      longitude    specificity       vicinity          crit1          crit2          crit3      doubtterr   alternative       multiple        success        suicide    attacktype1  attacktype2  attacktype3      targtype1   targsubtype1        natlty1     targtype2  targsubtype2       natlty2    targtype3  targsubtype3      natlty3    guncertain1  guncertain2  guncertain3     individual         nperps       nperpcap        claimed     claimmode       claim2  claimmode2      claim3  claimmode3    compclaim      weaptype1   weapsubtype1     weaptype2  weapsubtype2    weaptype3  weapsubtype3  weaptype4  weapsubtype4          nkill        nkillus       nkillter         nwound       nwoundus       nwoundte       property    propextent     propvalue      ishostkid      nhostkid    nhostkidus       nhours         ndays        ransom     ransomamt   ransomamtus    ransompaid  ransompaidus  hostkidoutcome     nreleased        INT_LOG       INT_IDEO       INT_MISC        INT_ANY
count  2.097060e+05  209706.000000  209706.000000  209706.000000  209706.000000  209706.000000  209706.000000  205015.000000  205014.000000  209705.000000  209706.000000  209706.000000  209706.000000  209706.000000  209706.000000  35249.000000  209705.000000  209706.000000  209706.000000  209706.000000  8434.000000   658.000000  209706.000000  197867.000000  207692.000000  14535.000000  14003.000000  14211.000000  1629.000000   1539.000000  1600.000000  209326.000000  2538.000000   410.000000  209706.000000  138613.000000  140233.000000  143613.000000  24696.000000  2476.000000  732.000000  409.000000  140.000000  4954.000000  209706.000000  183765.000000  16402.000000  14292.000000  2371.000000   2132.000000  73.000000     70.000000  197179.000000  145269.000000  141547.000000  189770.000000  145009.000000  138800.000000  209706.000000  73464.000000  4.837500e+04  209528.000000  16667.000000  16612.000000  4985.000000  10301.000000  80466.000000  1.533000e+03  7.340000e+02  9.510000e+02    725.000000    14091.000000  13494.000000  209706.000000  209706.000000  209706.000000  209706.000000
mean   2.004867e+11    2004.800993       6.455285      15.527930       0.051525     130.291351       7.279854      23.358696      30.416738       1.468387       0.068854       0.988265       0.993291       0.866246      -0.423593      1.293001       0.145900       0.883628       0.035469       3.351311     3.956011     5.705167       8.484078      47.163034     127.079035     10.050705     54.641720    126.196749    10.171885     55.948668   132.108125       0.075949     0.258077     0.185366       0.003929     -68.819613      -1.341261       0.079095      7.069768     0.235864    7.312842    0.337408    6.692857    -6.131005       6.543060      11.081294      6.986404     11.206969     7.138338     12.407598   6.246575     10.842857       2.431030       0.038797       0.559122       3.085872       0.034177       0.137759      -0.712927      3.289407  1.750017e+05       0.052967      3.040559     -0.311281   -55.121916    -36.012717     -0.157520  2.791526e+06  3.211036e+05  6.204108e+05    182.750345        4.705912    -31.945531      -4.424289      -4.339318       0.086841      -3.850758
std    1.351933e+09      13.519321       3.387098       8.801104       0.221066     111.714562       2.905697      18.137061      56.113029       0.984958       0.281308       0.107693       0.081636       0.340389       2.305025      0.733146       0.353007       0.320672       0.184962       2.029153     2.358046     2.089622       6.602032      30.599430      91.732261      5.732859     25.761594    122.995161     5.649132     25.870299   152.871895       0.264916     0.437663     0.389069       0.062561     194.556994      12.115647       0.997974      2.343068     0.872997    2.683832    0.473405    2.853488     4.318717       2.296739       6.485577      2.377310      7.727262

- Since, there are many columns and the naming convention is also weird, we will have to first rename the columns.
- But before that let us check for the presence of null values in our data.

In [6]:
df.isna().sum()

eventid            0
iyear              0
imonth             0
iday               0
approxdate    197017
               ...  
INT_LOG            0
INT_IDEO           0
INT_MISC           0
INT_ANY            0
related       179102
Length: 135, dtype: int64

- Let us have a closer look at the above result and to ease our observation, let us convert the above into a dataframe.

In [7]:
pd.set_option('display.max_rows',None)  # We want to see all the rows of our columns_df
miss_info = df.isna().sum()
columns_df = miss_info.to_frame()   # converting miss_info to a dataframe

columns_df.rename(columns={0:"Missing values"}, inplace=True)   

high_missing_cols = columns_df[columns_df['Missing values'] > 0.50*df.shape[0]]

- high_missing_cols shows the columns with more than 50% missing values.
- Since, these columns have high number of missing values and it would make no sense to impute them (for missing value removal), we would drop them from our dataframe.

In [8]:
high_missing_cols

Missing values
approxdate                  197017
resolution                  204845
location                    142231
alternative                 174457
alternative_txt             174457
attacktype2                 201272
attacktype2_txt             201272
attacktype3                 209048
attacktype3_txt             209048
targtype2                   195171
targtype2_txt               195171
targsubtype2                195703
targsubtype2_txt            195703
corp2                       196198
target2                     195296
natlty2                     195495
natlty2_txt                 195495
targtype3                   208077
targtype3_txt               208077
targsubtype3                208167
targsubtype3_txt            208167
corp3                       208228
target3                     208078
natlty3                     208106
natlty3_txt                 208106
gsubname                    202476
gname2                      207127
gsubname2                   209522
gname3                      209292
gsubname3                   209683
motive                      154648
guncertain2                 207168
guncertain3                 209296
claimmode                   185010
claimmode_txt               185010
claim2                      207230
claimmode2                  208974
claimmode2_txt              208974
claim3                      209297
claimmode3                  209566
claimmode3_txt              209566
compclaim                   204752
weaptype2                   193304
weaptype2_txt               193304
weapsubtype2                195414
weapsubtype2_txt            195414
weaptype3                   207335
weaptype3_txt               207335
weapsubtype3                207574
weapsubtype3_txt            207574
weaptype4                   209633
weaptype4_txt               209633
weapsubtype4                209636
weapsubtype4_txt            209636
propextent                  136242
propextent_txt              136242
propvalue                   161331
propcomment                 142282
nhostkid                    193039
nhostkidus                  193094
nhours                      204721
ndays                       199405
divert                      209368
kidhijcountry               206393
ransom                      129240
ransomamt                   208173
ransomamtus                 208972
ransompaid                  208755
ransompaidus                208981
ransomnote                  209131
hostkidoutcome              195615
hostkidoutcome_txt          195615
nreleased                   196212
addnotes                    173897
scite2                      116963
scite3                      157030
related                     179102

In [9]:
print(f'Basically we have {len(high_missing_cols)} columns with more than 50% missing values')

Basically we have 77 columns with more than 50% missing values


- Next, we drop these columns with high missing values.

In [10]:
df_copy = df.copy() # saving the copy of original data, just in case

In [11]:
df.drop(columns=list(high_missing_cols.index), inplace=True)   # taking list of missing values of cols and dropping them from df

In [12]:
df.shape      # Checking the shape of the dataframe after dropping high missing value cols

(209706, 58)

- Let us recheck the number of missing values in each column after column removal in previous step

In [13]:
(df.isna().sum()/len(df)*100).sort_values(ascending=False)

weapdetail          41.281127
nperps              33.901271
nwoundte            33.812099
nperpcap            33.128761
nkillter            32.502170
scite1              31.559421
summary             31.529856
claimed             31.516981
nwoundus            30.851287
nkillus             30.727304
corp1               20.282681
weapsubtype1        12.370175
weapsubtype1_txt    12.370175
nwound               9.506643
nkill                5.973601
targsubtype1_txt     5.645523
targsubtype1         5.645523
longitude            2.237418
latitude             2.236941
natlty1              0.960392
natlty1_txt          0.960392
target1              0.302805
city                 0.203142
guncertain1          0.181206
ishostkid            0.084881
multiple             0.000477
specificity          0.000477
dbsource             0.000000
individual           0.000000
property             0.000000
weaptype1_txt        0.000000
weaptype1            0.000000
INT_LOG              0.000000
INT_IDEO  

- Still, there are columns with high missing values and there are some columns with no missing values.
- Let us dive deeper into the columns with **high** missing values to understand them better and decide on keeping them.

In [14]:
len(df['weapdetail'].value_counts())

21362

In [15]:
df['weapdetail'].value_counts()[:10]

Explosive                                                 20918
Automatic firearm                                         12419
A roadside bomb was used in the attack.                    4611
Pistol                                                     4472
Incendiary                                                 4152
Firearm                                                    4053
Unknown firearms were used in the attack.                  4028
An improvised explosive device was used in the attack.     3024
Mortars were used in the attack.                           1083
Unknown explosives were used in the attack.                1072
Name: weapdetail, dtype: int64

- 'weapdetail' column has *too many unique categories* which means high variance alongwith no pattern in details, except that **majority** of the values do not belong to a particular category but are rather varied.
- So, we are going to drop this column as it has **too many missing values (41%)** also.

In [16]:
df['nperps'].value_counts()[:10]

-99.0    106061
 1.0       9968
 2.0       6939
 3.0       3184
 4.0       2473
 5.0       1259
 6.0        881
 10.0       820
Name: nperps, dtype: int64

- '-99' means that number of perpetrators is **"Unknown"** (given in column description) and they are around **50%** in the column.
- Column has around **34%** missing values too, so it shows that around **50+34 = 84%** of data in "nperps" column has either *unknown* or *missing* values.
- Thus, we will drop this column.

In [17]:
(df['nwoundte'].value_counts())[:10]

0.0     135068
1.0       1159
2.0        607
3.0        440
4.0        290
5.0        225
6.0        170
7.0        136
10.0       105
Name: nwoundte, dtype: int64

- Replace the missing values in this column with '-1' as in majority of cases, “Number of Perpetrator Fatalities” (nwoundte) is 0, but in some cases there are high fatalities.
- Look more into cases, where there were these fatalities.
- Also, change datatype of this column from "float" to "integer".

In [18]:
(df['nperpcap'].value_counts())[:10]

 0.0     132809
 1.0       2711
-99.0      2073
 2.0       1021
 3.0        554
 4.0        331
 5.0        184
 6.0        109
 8.0         64
 7.0         55
 10.0        48
Name: nperpcap, dtype: int64

- In majority of cases, perpetrators were not captured and there are missing values too. 
- Mark the missing values as '-1' and look more into the ones where perpetrators were taken into custody.
- Change the datatype, "float" to "integer".

In [19]:
(df['nkillter'].value_counts())[:10]

0.0     124842
1.0       7623
2.0       2451
3.0       1518
4.0        984
5.0        734
6.0        545
7.0        374
8.0        346
10.0       323
Name: nkillter, dtype: int64

- Since, this column also follows similar convention to 'nwoundte' column (given in codebook), drop this column.

In [20]:
(df['scite1'].value_counts())[:5]

Committee on Government Operations United States Senate, "Riots, Civil, and Criminal Disorders," U.S. Government Printing Office, August 6, 1970.    205
"Afghan War Casualty Report: October 2019," NYTimes.com Feed, October 4, 2019.                                                                       116
"Afghan War Casualty Report: September 2020," NYTimes.com Feed, September 3, 2020.                                                                   107
"ISIS detonates 5 empty homes in Daur using IEDs," Iraqi News, September 28, 2016.                                                                   106
"Afghan War Casualty Report: March 2020," NYTimes.com Feed, March 5, 2020.                                                                            92
Name: scite1, dtype: int64

- 'scite1' cites the first source that was used to compile information on the specific incident.
- Not very relevant for our analysis, dropping it.

In [21]:
(df['summary'].value_counts())[:4]

09/00/2016: Sometime between September 18, 2016 and September 24, 2016, assailants detonated an explosive device at a civilian house in Dawr, Saladin, Iraq. There were no reported casualties resulting from the blast. No group claimed responsibility for the incident; however, sources attributed the attack to the Islamic State of Iraq and the Levant (ISIL).    100
12/01/2016: Assailants detonated explosive devices that destroyed an electricity transmission tower in Albu Bali, Al Anbar, Iraq. This was one of 80 similar attacks targeting towers that resulted in the death of two military engineering team members. No group claimed responsibility for the incidents.                                                             80
05/29/2016: Security forces discovered and defused an explosives-rigged house in Sejar, Al Anbar, Iraq. No group claimed responsibility for the incident.                                                                                                                     

- 'summary' gives a brief summary of the attack.
- Not very relevant to our analysis, as it is not available for all the records as well as it has sentences as its input.
- Drop the column.

In [22]:
df['claimed'].value_counts()

 0.0    117524
 1.0     24616
-9.0      1473
Name: claimed, dtype: int64

In [23]:
df['claimed'].isna().sum()

66093

- Majority of attacks are not claimed by any person or group.
- Mark the ones where the values is "unknown" as "-9" as it represent the one which are unknown and we have no idea on whether missing value attacks were claimed or not.
- Change datatype of column to "integer".

In [24]:
print((df['nwoundus'].value_counts())[:10])
print('')
print((df['nkillus'].value_counts())[:10])

0.0     144266
1.0        352
2.0        124
3.0         75
4.0         41
5.0         22
9.0         15
7.0         15
10.0        14
Name: nwoundus, dtype: int64

0.0     144295
1.0        634
2.0        171
3.0         56
4.0         38
5.0         20
6.0         12
7.0          4
17.0         4
15.0         3
9.0          3
10.0         3
Name: nkillus, dtype: int64


- Replace missing values "-9" or "-99" as convention for all missing values in the previous columns (or decide one)
- change the datatype to "integer"
- Rename the columns too.

In [25]:
len(df['corp1'].value_counts())

38368

- Since, there are many names of the corporate entity or government agency that were targeted, we will keep this column and further look into the variable. 
- Impute the missing values with "Unknown"

In [26]:
print(df['weapsubtype1_txt'].value_counts()[:5])
print('')
df[['weapsubtype1', 'weapsubtype1_txt']].isna().sum()

Unknown Explosive Type                       48807
Unknown Gun Type                             40783
Automatic or Semi-Automatic Rifle            18176
Projectile (rockets, mortars, RPGs, etc.)    12031
Vehicle                                      10849
Name: weapsubtype1_txt, dtype: int64



weapsubtype1        25941
weapsubtype1_txt    25941
dtype: int64

- Since, both 'weapsubtype1' and 'weapsubtype1_txt' are categorical variables and represent same information, we will keep only one of them, i.e., 'weapsubtype1_txt' as it gives textual information of weapon subtype.
- Impute missing values with **mode** of the column.
- Rename the column too.

In [27]:
print((df['nwound'].value_counts())[:10])
print('')
print((df['nwoundus'].value_counts())[:10])

0.0     116963
1.0      19036
2.0      12432
3.0       8828
4.0       5775
5.0       4440
6.0       3296
7.0       2740
8.0       2046
10.0      1604
Name: nwound, dtype: int64

0.0     144266
1.0        352
2.0        124
3.0         75
4.0         41
5.0         22
9.0         15
7.0         15
10.0        14
Name: nwoundus, dtype: int64


- For both the columns, impute the missing values with "mode"
- Rename the columns and also change the datatype.

- We would repeat similar strategies (as discussed above) with the remaining columns, and perform missing values imputation accordingly.
- Also, we would drop the columns which have similar counterparts which impart the same information.
- Remaining steps would be described as we proceed.

<h1  style="text-align: center"  class="list-group-item list-group-item-warning"> <span style="color:#2E6AA9">Data Wrangling</span>  </h1> 

Now, let us start dropping the columns which we decided and also perform missing value imputation, change datatypes of required columns and rename the columns. Before that, let us have a look at columns with missing values.

In [28]:
(df.isna().sum()/len(df)*100).sort_values(ascending=False)

weapdetail          41.281127
nperps              33.901271
nwoundte            33.812099
nperpcap            33.128761
nkillter            32.502170
scite1              31.559421
summary             31.529856
claimed             31.516981
nwoundus            30.851287
nkillus             30.727304
corp1               20.282681
weapsubtype1        12.370175
weapsubtype1_txt    12.370175
nwound               9.506643
nkill                5.973601
targsubtype1_txt     5.645523
targsubtype1         5.645523
longitude            2.237418
latitude             2.236941
natlty1              0.960392
natlty1_txt          0.960392
target1              0.302805
city                 0.203142
guncertain1          0.181206
ishostkid            0.084881
multiple             0.000477
specificity          0.000477
dbsource             0.000000
individual           0.000000
property             0.000000
weaptype1_txt        0.000000
weaptype1            0.000000
INT_LOG              0.000000
INT_IDEO  

In [29]:
cols_to_drop = ['nperps', 'weapdetail', 'scite1', 'summary', 'weapsubtype1', 'targsubtype1', 
                'natlty1', 'weaptype1', 'eventid', 'country', 'region', 'attacktype1', 'targtype1', 'specificity', 'individual']    

df.drop(columns=cols_to_drop, inplace=True)

In [30]:
df.head()

iyear  imonth  iday  extended         country_txt                   region_txt provstate           city   latitude   longitude  vicinity  crit1  crit2  crit3  doubtterr  multiple  success  suicide                 attacktype1_txt                targtype1_txt                                      targsubtype1_txt                        corp1                  target1         natlty1_txt                               gname  guncertain1  nperpcap  claimed weaptype1_txt        weapsubtype1_txt  nkill  nkillus  nkillter  nwound  nwoundus  nwoundte  property  ishostkid dbsource  INT_LOG  INT_IDEO  INT_MISC  INT_ANY
0   1970       7     2         0  Dominican Republic  Central America & Caribbean  National  Santo Domingo  18.456792  -69.951164         0      1      1      1          0       0.0        1        0                   Assassination  Private Citizens & Property                                        Named Civilian                          NaN             Julio Guzman  Dominican Republic                              MANO-D          0.0       NaN      NaN       Unknown                     NaN    1.0      NaN       NaN     0.0       NaN       NaN         0        0.0     PGIS        0         0         0        0
1   1970       0     0         0              Mexico                North America   Federal    Mexico city  19.371887  -99.086624         0      1      1      1          0       0.0        1        0     Hostage Taking (Kidnapping)      Government (Diplomatic)  Diplomatic Personnel (outside of embassy, consulate)  Belgian Ambassador Daughter  Nadine Chaval, daughter             Belgium  23rd of September Communist League          0.0       NaN      NaN       Unknown                     NaN    0.0      NaN       NaN     0.0       NaN       NaN         0        1.0     PGIS        0         1         1        1
2   1970       1     0         0         Philippines               Southeast Asia    Tarlac        Unknown  15.478598  120.599741         0      1      1      1          0       0.0        1        0                   Assassination          Journalists & Media                       Radio Journalist/Staff/Facility             Voice of America                 Employee       United States                             Unknown          0.0       NaN      NaN       Unknown                     NaN    1.0      NaN       NaN     0.0       NaN       NaN         0        0.0     PGIS       -9        -9         1        1
3   1970       1     0         0              Greece               Western Europe    Attica         Athens  37.997490   23.762728         0      1      1      1          0       0.0        1        0               Bombing/Explosion      Government (Diplomatic)                                     Embassy/Consulate                          NaN             U.S. Embassy       United States                             Unknown          0.0       NaN      NaN    Explosives  Unknown Explosive Type    NaN      NaN       NaN     NaN       NaN       NaN         1        0.0     PGIS       -9        -9         1        1
4   1970       1     0         0               Japan                    East Asia   Fukouka        Fukouka  33.580412  130.396361         0      1      1      1         -9       0.0        1        0  Facility/Infrastructure Attack      Government (Diplomatic)                                     Embassy/Consulate                          NaN           U.S. Consulate       United States                             Unknown          0.0       NaN      NaN    Incendiary                     NaN    NaN      NaN       NaN     NaN       NaN       NaN         1        0.0     PGIS       -9        -9         1        1

In [31]:
df.shape

(209706, 43)

In [32]:
(df.isna().sum()/len(df)*100).sort_values(ascending=False)

nwoundte            33.812099
nperpcap            33.128761
nkillter            32.502170
claimed             31.516981
nwoundus            30.851287
nkillus             30.727304
corp1               20.282681
weapsubtype1_txt    12.370175
nwound               9.506643
nkill                5.973601
targsubtype1_txt     5.645523
longitude            2.237418
latitude             2.236941
natlty1_txt          0.960392
target1              0.302805
city                 0.203142
guncertain1          0.181206
ishostkid            0.084881
multiple             0.000477
weaptype1_txt        0.000000
INT_IDEO             0.000000
INT_MISC             0.000000
INT_LOG              0.000000
property             0.000000
dbsource             0.000000
iyear                0.000000
gname                0.000000
crit1                0.000000
iday                 0.000000
extended             0.000000
country_txt          0.000000
region_txt           0.000000
provstate            0.000000
vicinity  

Let us perform missing value imputation as well as datatype change of columns. 
- Before that, let us **change datatype** of float columns to integer. 

In [33]:
float_columns = [col for col in df.columns if df[col].dtype == 'float64']
float_columns

['latitude',
 'longitude',
 'multiple',
 'guncertain1',
 'nperpcap',
 'claimed',
 'nkill',
 'nkillus',
 'nkillter',
 'nwound',
 'nwoundus',
 'nwoundte',
 'ishostkid']

- Except, "latitude" and "longitude" column, we would change datatype of all the "float" columns to "integer"
- But first we will have to impute **missing values** otherwise the code will throw an error.

In [34]:
## Perfoming missing value imputation on all the columns at once

df['multiple'].fillna(1, inplace=True)
df['latitude'].fillna(df['latitude'].mode()[0], inplace=True)  # As the missing values are comparatively low, we are using mode to fill them
df['longitude'].fillna(df['longitude'].mode()[0], inplace=True)
df['guncertain1'].fillna(df['guncertain1'].mode()[0], inplace=True)

df['nperpcap'].fillna(-99.0, inplace=True)                # As seen earlier, "-99.0" is being used to signify unknown (missing value) 
df['claimed'].fillna(-99.0, inplace=True)                 # In our data
df['nkill'].fillna(-99.0, inplace=True)
df['nkillus'].fillna(-99.0, inplace=True)
df['nkillter'].fillna(-99.0, inplace=True)
df['nwound'].fillna(-99.0, inplace=True)
df['nwoundus'].fillna(-99.0, inplace=True)
df['nwoundte'].fillna(-99.0, inplace=True)
df['ishostkid'].fillna(-9.0, inplace=True)

In [35]:
df['nperpcap'].replace(-9.0, -99.0, inplace=True) # replacing -9.0 (very small in number) with -99.0 as unknown
df['claimed'].replace(-9.0, -99.0, inplace=True) 

In [36]:
df['city'].fillna('Missing Info', inplace=True)        
df['targsubtype1_txt'].fillna('Missing Info', inplace=True)
df['corp1'].fillna('Missing Info', inplace=True)
df['target1'].fillna('Missing Info', inplace=True)
df['natlty1_txt'].fillna('Missing Info', inplace=True)
df['weapsubtype1_txt'].fillna('Missing Info', inplace=True)

- As the columns above also have "unknown" values, we have therefore represented missing values with **"Missing Info"** to avoid any confusion.

In [37]:
df.isna().sum()      ## Checking for missing values

iyear               0
imonth              0
iday                0
extended            0
country_txt         0
region_txt          0
provstate           0
city                0
latitude            0
longitude           0
vicinity            0
crit1               0
crit2               0
crit3               0
doubtterr           0
multiple            0
success             0
suicide             0
attacktype1_txt     0
targtype1_txt       0
targsubtype1_txt    0
corp1               0
target1             0
natlty1_txt         0
gname               0
guncertain1         0
nperpcap            0
claimed             0
weaptype1_txt       0
weapsubtype1_txt    0
nkill               0
nkillus             0
nkillter            0
nwound              0
nwoundus            0
nwoundte            0
property            0
ishostkid           0
dbsource            0
INT_LOG             0
INT_IDEO            0
INT_MISC            0
INT_ANY             0
dtype: int64

- No Missing values in data now.
- But we need to **change Datatypes** of many columns as well as rename them.

In [38]:
df.dtypes

iyear                 int64
imonth                int64
iday                  int64
extended              int64
country_txt          object
region_txt           object
provstate            object
city                 object
latitude            float64
longitude           float64
vicinity              int64
crit1                 int64
crit2                 int64
crit3                 int64
doubtterr             int64
multiple            float64
success               int64
suicide               int64
attacktype1_txt      object
targtype1_txt        object
targsubtype1_txt     object
corp1                object
target1              object
natlty1_txt          object
gname                object
guncertain1         float64
nperpcap            float64
claimed             float64
weaptype1_txt        object
weapsubtype1_txt     object
nkill               float64
nkillus             float64
nkillter            float64
nwound              float64
nwoundus            float64
nwoundte            

In [39]:
float_int_columns = [col for col in df.columns if df[col].dtype == 'float64' or 'int64']
print(float_int_columns)

['iyear', 'imonth', 'iday', 'extended', 'country_txt', 'region_txt', 'provstate', 'city', 'latitude', 'longitude', 'vicinity', 'crit1', 'crit2', 'crit3', 'doubtterr', 'multiple', 'success', 'suicide', 'attacktype1_txt', 'targtype1_txt', 'targsubtype1_txt', 'corp1', 'target1', 'natlty1_txt', 'gname', 'guncertain1', 'nperpcap', 'claimed', 'weaptype1_txt', 'weapsubtype1_txt', 'nkill', 'nkillus', 'nkillter', 'nwound', 'nwoundus', 'nwoundte', 'property', 'ishostkid', 'dbsource', 'INT_LOG', 'INT_IDEO', 'INT_MISC', 'INT_ANY']


In [40]:
df[float_int_columns].head()    ## Having a look at float and int columns

iyear  imonth  iday  extended         country_txt                   region_txt provstate           city   latitude   longitude  vicinity  crit1  crit2  crit3  doubtterr  multiple  success  suicide                 attacktype1_txt                targtype1_txt                                      targsubtype1_txt                        corp1                  target1         natlty1_txt                               gname  guncertain1  nperpcap  claimed weaptype1_txt        weapsubtype1_txt  nkill  nkillus  nkillter  nwound  nwoundus  nwoundte  property  ishostkid dbsource  INT_LOG  INT_IDEO  INT_MISC  INT_ANY
0   1970       7     2         0  Dominican Republic  Central America & Caribbean  National  Santo Domingo  18.456792  -69.951164         0      1      1      1          0       0.0        1        0                   Assassination  Private Citizens & Property                                        Named Civilian                 Missing Info             Julio Guzman  Dominican Republic                              MANO-D          0.0     -99.0    -99.0       Unknown            Missing Info    1.0    -99.0     -99.0     0.0     -99.0     -99.0         0        0.0     PGIS        0         0         0        0
1   1970       0     0         0              Mexico                North America   Federal    Mexico city  19.371887  -99.086624         0      1      1      1          0       0.0        1        0     Hostage Taking (Kidnapping)      Government (Diplomatic)  Diplomatic Personnel (outside of embassy, consulate)  Belgian Ambassador Daughter  Nadine Chaval, daughter             Belgium  23rd of September Communist League          0.0     -99.0    -99.0       Unknown            Missing Info    0.0    -99.0     -99.0     0.0     -99.0     -99.0         0        1.0     PGIS        0         1         1        1
2   1970       1     0         0         Philippines               Southeast Asia    Tarlac        Unknown  15.478598  120.599741         0      1      1      1          0       0.0        1        0                   Assassination          Journalists & Media                       Radio Journalist/Staff/Facility             Voice of America                 Employee       United States                             Unknown          0.0     -99.0    -99.0       Unknown            Missing Info    1.0    -99.0     -99.0     0.0     -99.0     -99.0         0        0.0     PGIS       -9        -9         1        1
3   1970       1     0         0              Greece               Western Europe    Attica         Athens  37.997490   23.762728         0      1      1      1          0       0.0        1        0               Bombing/Explosion      Government (Diplomatic)                                     Embassy/Consulate                 Missing Info             U.S. Embassy       United States                             Unknown          0.0     -99.0    -99.0    Explosives  Unknown Explosive Type  -99.0    -99.0     -99.0   -99.0     -99.0     -99.0         1        0.0     PGIS       -9        -9         1        1
4   1970       1     0         0               Japan                    East Asia   Fukouka        Fukouka  33.580412  130.396361         0      1      1      1         -9       0.0        1        0  Facility/Infrastructure Attack      Government (Diplomatic)                                     Embassy/Consulate                 Missing Info           U.S. Consulate       United States                             Unknown          0.0     -99.0    -99.0    Incendiary            Missing Info  -99.0    -99.0     -99.0   -99.0     -99.0     -99.0         1        0.0     PGIS       -9        -9         1        1

- We are not going to change datatypes of **"latitude"** and **"longitude"** columns
- But many float and integer columns are actually **Categorical""** variables, which need to be mapped to their respective "Yes", "No" or "Unknown" category values.
- Separately change their datatype to "category" alongwith making required float columns into integer ones.

In [41]:
## The columns below are actually integer numeric columns but are in wrong datatype, i.e., float
float_to_integer = ['nperpcap', 'nkill', 'nkillus', 'nkillter', 'nwound', 'nwoundus', 'nwoundte']

df[float_to_integer] = df[float_to_integer].astype('int64')  ## converting the columns into integer datatype
df[float_to_integer].head()   # Reconfirming the changes

nperpcap  nkill  nkillus  nkillter  nwound  nwoundus  nwoundte
0       -99      1      -99       -99       0       -99       -99
1       -99      0      -99       -99       0       -99       -99
2       -99      1      -99       -99       0       -99       -99
3       -99    -99      -99       -99     -99       -99       -99
4       -99    -99      -99       -99     -99       -99       -99

- We are done with converting the datatype from float to integer of required columns
- Now, let us map the values of remaining (required) float and integer columns to their respective categorical values.

In [42]:
df.dtypes

iyear                 int64
imonth                int64
iday                  int64
extended              int64
country_txt          object
region_txt           object
provstate            object
city                 object
latitude            float64
longitude           float64
vicinity              int64
crit1                 int64
crit2                 int64
crit3                 int64
doubtterr             int64
multiple            float64
success               int64
suicide               int64
attacktype1_txt      object
targtype1_txt        object
targsubtype1_txt     object
corp1                object
target1              object
natlty1_txt          object
gname                object
guncertain1         float64
nperpcap              int64
claimed             float64
weaptype1_txt        object
weapsubtype1_txt     object
nkill                 int64
nkillus               int64
nkillter              int64
nwound                int64
nwoundus              int64
nwoundte            

In [43]:
float_cat_cols = ['extended', 'vicinity', 'crit1', 'crit2', 'crit3', 'doubtterr', 'multiple', 'success', 'suicide', 'guncertain1', 
                  'claimed', 'property', 'ishostkid', 'INT_LOG', 'INT_IDEO', 'INT_MISC', 'INT_ANY']

# The columns above would be mapped to their respective categorical values

In [44]:
df['extended'].value_counts()

0    198901
1     10805
Name: extended, dtype: int64

In [45]:
df['extended'] = df['extended'].map({1:"Yes", 0:"No"}).astype('object')
df['vicinity'] = df['vicinity'].map({1:"Yes", 0:"No", -9:"unknown"}).astype('object')
df['crit1'] = df['crit1'].map({1:"Yes", 0:"No"}).astype('object')
df['crit2'] = df['crit2'].map({1:"Yes", 0:"No"}).astype('object')
df['crit3'] = df['crit3'].map({1:"Yes", 0:"No"}).astype('object')
df['doubtterr'] = df['doubtterr'].map({1:"Yes", 0:"No", -9:"Variable omitted during data collection"}).astype('object')
df['multiple'] = df['multiple'].map({1.0:"Yes", 0.0:"No"}).astype('object')
df['success'] = df['success'].map({1:"Yes", 0:"No"}).astype('object')
df['suicide'] = df['suicide'].map({1:"Yes", 0:"No"}).astype('object')
df['guncertain1'] = df['guncertain1'].map({1.0:"Yes", 0.0:"No"}).astype('object')
df['claimed'] = df['claimed'].map({1.0:"Yes", 0.0:"No", -99.0:"Unknown"}).astype('object')
df['property'] = df['property'].map({1:"Yes", 0:"No", -9:"unknown"}).astype('object')
df['ishostkid'] = df['ishostkid'].map({1.0:"Yes", 0.0:"No", -9.0:"unknown"}).astype('object')

df['INT_LOG'] = df['INT_LOG'].map({1:"Yes", 0:"No", -9:"unknown"}).astype('object')
df['INT_IDEO'] = df['INT_IDEO'].map({1:"Yes", 0:"No", -9:"unknown"}).astype('object')
df['INT_MISC'] = df['INT_MISC'].map({1:"Yes", 0:"No", -9:"unknown"}).astype('object')
df['INT_ANY'] = df['INT_ANY'].map({1:"Yes", 0:"No", -9:"unknown"}).astype('object')

- We have used the codebook.pdf document to map the integer values to the respective string values
- Alongwith mapping, we have also converted the datatype of the mapped columns
- Finally, we are left with renaming the columns


**NOTE**: For the numeric columns, we have imputed missing values with "-99" which just points it out as missing value for the respective column.

In [46]:
df.dtypes

iyear                 int64
imonth                int64
iday                  int64
extended             object
country_txt          object
region_txt           object
provstate            object
city                 object
latitude            float64
longitude           float64
vicinity             object
crit1                object
crit2                object
crit3                object
doubtterr            object
multiple             object
success              object
suicide              object
attacktype1_txt      object
targtype1_txt        object
targsubtype1_txt     object
corp1                object
target1              object
natlty1_txt          object
gname                object
guncertain1          object
nperpcap              int64
claimed              object
weaptype1_txt        object
weapsubtype1_txt     object
nkill                 int64
nkillus               int64
nkillter              int64
nwound                int64
nwoundus              int64
nwoundte            

In [47]:
df.head()

iyear  imonth  iday extended         country_txt                   region_txt provstate           city   latitude   longitude vicinity crit1 crit2 crit3                                doubtterr multiple success suicide                 attacktype1_txt                targtype1_txt                                      targsubtype1_txt                        corp1                  target1         natlty1_txt                               gname guncertain1  nperpcap  claimed weaptype1_txt        weapsubtype1_txt  nkill  nkillus  nkillter  nwound  nwoundus  nwoundte property ishostkid dbsource  INT_LOG INT_IDEO INT_MISC INT_ANY
0   1970       7     2       No  Dominican Republic  Central America & Caribbean  National  Santo Domingo  18.456792  -69.951164       No   Yes   Yes   Yes                                       No       No     Yes      No                   Assassination  Private Citizens & Property                                        Named Civilian                 Missing Info             Julio Guzman  Dominican Republic                              MANO-D          No       -99  Unknown       Unknown            Missing Info      1      -99       -99       0       -99       -99       No        No     PGIS       No       No       No      No
1   1970       0     0       No              Mexico                North America   Federal    Mexico city  19.371887  -99.086624       No   Yes   Yes   Yes                                       No       No     Yes      No     Hostage Taking (Kidnapping)      Government (Diplomatic)  Diplomatic Personnel (outside of embassy, consulate)  Belgian Ambassador Daughter  Nadine Chaval, daughter             Belgium  23rd of September Communist League          No       -99  Unknown       Unknown            Missing Info      0      -99       -99       0       -99       -99       No       Yes     PGIS       No      Yes      Yes     Yes
2   1970       1     0       No         Philippines               Southeast Asia    Tarlac        Unknown  15.478598  120.599741       No   Yes   Yes   Yes                                       No       No     Yes      No                   Assassination          Journalists & Media                       Radio Journalist/Staff/Facility             Voice of America                 Employee       United States                             Unknown          No       -99  Unknown       Unknown            Missing Info      1      -99       -99       0       -99       -99       No        No     PGIS  unknown  unknown      Yes     Yes
3   1970       1     0       No              Greece               Western Europe    Attica         Athens  37.997490   23.762728       No   Yes   Yes   Yes                                       No       No     Yes      No               Bombing/Explosion      Government (Diplomatic)                                     Embassy/Consulate                 Missing Info             U.S. Embassy       United States                             Unknown          No       -99  Unknown    Explosives  Unknown Explosive Type    -99      -99       -99     -99       -99       -99      Yes        No     PGIS  unknown  unknown      Yes     Yes
4   1970       1     0       No               Japan                    East Asia   Fukouka        Fukouka  33.580412  130.396361       No   Yes   Yes   Yes  Variable omitted during data collection       No     Yes      No  Facility/Infrastructure Attack      Government (Diplomatic)                                     Embassy/Consulate                 Missing Info           U.S. Consulate       United States                             Unknown          No       -99  Unknown    Incendiary            Missing Info    -99      -99       -99     -99       -99       -99      Yes        No     PGIS  unknown  unknown      Yes     Yes

- Finally, let us **rename our columns** with better names **to avoid confusion and provide more clarity** about the variable.
- We would be using codebook.pdf (provide by University of Maryland) for reference.

In [48]:
df.rename(columns={'iyear':'Year','imonth':'Month','iday':'Day','extended':'Extended Incident?','country_txt':'Country',
                  'region_txt':'Region','provstate':'Province / Administrative Region / State','city':'City','latitude':'Latitude',
                  'longitude':'Longitude','vicinity':'Vicinity','crit1':'POLITICAL, ECONOMIC, RELIGIOUS, OR SOCIAL GOAL',
                  'crit2':'INTENTION TO COERCE, INTIMIDATE OR PUBLICIZE TO LARGER AUDIENCE(S)','crit3':'OUTSIDE INTERNATIONAL HUMANITARIAN LAW',
                  'doubtterr':'Doubt if Terrorism?','multiple':'Part of Multiple Incident', 'success':'Successful Attack','suicide':'Suicide Attack',
                  'attacktype1_txt':'Attack Type','targtype1_txt':'Target/Victim Type','targsubtype1_txt':'Target/Victim Subtype',
                  'corp1':'Name of Entity','target1':'Specific Target/Victim','natlty1_txt':'Nationality of Target/Victim','gname':'Perpetrator Group Name',
                  'guncertain1':'First Perpetrator Group Suspected/Unconfirmed?','nperpcap':'Number of Perpetrators Captured',
                  'claimed':'Claim of Responsibility?','weaptype1_txt':'Weapon Type','weapsubtype1_txt':'Weapon Sub-type',
                  'nkill':'Total Number of Fatalities','nkillus':'Number of US Fatalities','nkillter':'Number of Perpetrator Fatalities',
                  'nwound':'Total Number of Injured','nwoundus':'Number of US Injured','nwoundte':'Number of Perpetrators Injured',
                  'property':'Property Damage','ishostkid':'Hostages or Kidnapping Victims','dbsource':'Data Collection',
                  'INT_LOG':'International- Logistical','INT_IDEO':'International- Ideological','INT_MISC':'International- Miscellaneous',
                  'INT_ANY':'International- Any of the above'}, inplace=True)

In [49]:
df.head()

Year  Month  Day Extended Incident?             Country                       Region Province / Administrative Region / State           City   Latitude   Longitude Vicinity POLITICAL, ECONOMIC, RELIGIOUS, OR SOCIAL GOAL INTENTION TO COERCE, INTIMIDATE OR PUBLICIZE TO LARGER AUDIENCE(S) OUTSIDE INTERNATIONAL HUMANITARIAN LAW                      Doubt if Terrorism? Part of Multiple Incident Successful Attack Suicide Attack                     Attack Type           Target/Victim Type                                 Target/Victim Subtype               Name of Entity   Specific Target/Victim Nationality of Target/Victim              Perpetrator Group Name First Perpetrator Group Suspected/Unconfirmed?  Number of Perpetrators Captured Claim of Responsibility? Weapon Type         Weapon Sub-type  Total Number of Fatalities  Number of US Fatalities  Number of Perpetrator Fatalities  Total Number of Injured  Number of US Injured  Number of Perpetrators Injured Property Damage Hostages or Kidnapping Victims Data Collection International- Logistical International- Ideological International- Miscellaneous International- Any of the above
0  1970      7    2                 No  Dominican Republic  Central America & Caribbean                                 National  Santo Domingo  18.456792  -69.951164       No                                            Yes                                                                Yes                                    Yes                                       No                        No               Yes             No                   Assassination  Private Citizens & Property                                        Named Civilian                 Missing Info             Julio Guzman           Dominican Republic                              MANO-D                                             No                              -99                  Unknown     Unknown            Missing Info                           1                      -99                               -99                        0                   -99                             -99              No                             No            PGIS                        No                         No                           No                              No
1  1970      0    0                 No              Mexico                North America                                  Federal    Mexico city  19.371887  -99.086624       No                                            Yes                                                                Yes                                    Yes                                       No                        No               Yes             No     Hostage Taking (Kidnapping)      Government (Diplomatic)  Diplomatic Personnel (outside of embassy, consulate)  Belgian Ambassador Daughter  Nadine Chaval, daughter                      Belgium  23rd of September Communist League                                             No                              -99                  Unknown     Unknown            Missing Info                           0                      -99                               -99                        0                   -99                             -99              No                            Yes            PGIS                        No                        Yes                          Yes                             Yes
2  1970      1    0                 No         Philippines               Southeast Asia                                   Tarlac        Unknown  15.478598  120.599741       No                                            Yes                                                                Yes                                    Yes                                       No                        No               Yes             No                   Assassination          Journalists & Media                       Radio Journalist/Staff/Facility             Voice 

- Let us map Month column values to their respective months.

<h1  style="text-align: center"  class="list-group-item list-group-item-warning"> <span style="color:#2E6AA9">Feature Engineering</span>  </h1> 

- We observed that we only have Month number in "Month" column, so we are going to build another column, i.e., "Name of Month", which would store name of the month of incident.
- Also, it would be better to create a column combining "Year", "Month" and "Day" which would lead to datetime column named "Date", which would help us in time series analysis.


- But we observed that there are "unknown" values in "month" column (around 20 values) and in "day" column (around 891 values)
- we will be using "Impute based on Frequency", where we can impute the 0s (the "unknown"s) with a value based on the frequency of each day or month. For example, we could impute each 0 with a day based on the proportion of occurrences of each day. Similar would be repeatde for month.
- This has to be done in order to make a datetime column feature, "Date" which would help in deeper analysis.

In [50]:
# Setting random seed for replicability
np.random.seed(123)

# Impute 0s based on day frequency
non_zero_days = df['Day'][(df['Day'] != 0) & (df['Day'] != 31) & (df['Day'] != 29)]    # To avoid leap year and every month has 30 in it
num_zeros = (df['Day'] == 0).sum()
imputed_values = np.random.choice(non_zero_days, num_zeros, replace=True)

# Replace the 0s with the randomly imputed values
df.loc[df['Day'] == 0, 'Day'] = np.random.choice(non_zero_days, num_zeros, replace=True)

# Convert the "Day" column back to integer type
df['Day'] = df['Day'].astype('int64')

# Repeating similar for Month column
df['Month'] = df['Month'].replace(0, df['Month'].mode()[0])

df['Month'] = df['Month'].astype('int64')

In [51]:
df['Name of Month'] = df['Month'].map({0:'unknown', 1:'January', 2:'February', 3:'March', 4:'April', 5:'May', 6:'June', 7:'July',
                              8:'August',9:'September',10:'October',11:'November', 12:'December'}).astype('object')

df['Date'] = pd.to_datetime(df[['Year','Month','Day']])

In [52]:
df.head()

Year  Month  Day Extended Incident?             Country                       Region Province / Administrative Region / State           City   Latitude   Longitude Vicinity POLITICAL, ECONOMIC, RELIGIOUS, OR SOCIAL GOAL INTENTION TO COERCE, INTIMIDATE OR PUBLICIZE TO LARGER AUDIENCE(S) OUTSIDE INTERNATIONAL HUMANITARIAN LAW                      Doubt if Terrorism? Part of Multiple Incident Successful Attack Suicide Attack                     Attack Type           Target/Victim Type                                 Target/Victim Subtype               Name of Entity   Specific Target/Victim Nationality of Target/Victim              Perpetrator Group Name First Perpetrator Group Suspected/Unconfirmed?  Number of Perpetrators Captured Claim of Responsibility? Weapon Type         Weapon Sub-type  Total Number of Fatalities  Number of US Fatalities  Number of Perpetrator Fatalities  Total Number of Injured  Number of US Injured  Number of Perpetrators Injured Property Damage Hostages or Kidnapping Victims Data Collection International- Logistical International- Ideological International- Miscellaneous International- Any of the above Name of Month       Date
0  1970      7    2                 No  Dominican Republic  Central America & Caribbean                                 National  Santo Domingo  18.456792  -69.951164       No                                            Yes                                                                Yes                                    Yes                                       No                        No               Yes             No                   Assassination  Private Citizens & Property                                        Named Civilian                 Missing Info             Julio Guzman           Dominican Republic                              MANO-D                                             No                              -99                  Unknown     Unknown            Missing Info                           1                      -99                               -99                        0                   -99                             -99              No                             No            PGIS                        No                         No                           No                              No          July 1970-07-02
1  1970      5   15                 No              Mexico                North America                                  Federal    Mexico city  19.371887  -99.086624       No                                            Yes                                                                Yes                                    Yes                                       No                        No               Yes             No     Hostage Taking (Kidnapping)      Government (Diplomatic)  Diplomatic Personnel (outside of embassy, consulate)  Belgian Ambassador Daughter  Nadine Chaval, daughter                      Belgium  23rd of September Communist League                                             No                              -99                  Unknown     Unknown            Missing Info                           0                      -99                               -99                        0                   -99                             -99              No                            Yes            PGIS                        No                        Yes                          Yes                             Yes           May 1970-05-15
2  1970      1   26                 No         Philippines               Southeast Asia                                   Tarlac        Unknown  15.478598  120.599741       No                                            Yes                                                                Yes                                    Yes                                       No                        No               Yes             No                   Assassination          Journalists & Med

In [54]:
## Final cleaned data

#df.to_csv('gtd_cleaned.csv', index=False)

# Summary of Data Cleaning

We cleaned the original dataset, i.e., `gtd.csv` and saved the cleaned dataset as an output csv file named `gtd_cleaned.csv`. Below are steps and approach we used in this notebook in concise manner:

## Dropping columns

- Checked for duplicate values (if any) in uncleaned data, fortunately, we had none.
- Next, we checked for missing values in our data and got a total missing values for each column.
- Dropped all the columns with high volume of missing values (more than **50%** missing values). Started with **135** columns, after dropping columns, ended up with **77** columns.
- `weapdetail` column had too many unique values alongwith high volumne of missing data (**41%**), so we dropped it too.
- `scite1` cited the first source that was used to compile information on the specific incident, it was not very relevant for our analysis, so dropped it. Similar goes with `summary` column.
- Both `weapsubtype1` and `weapsubtype1_txt` were categorical variables and represented same information, we only kept only one of them, i.e., `weapsubtype1_txt` as it gives textual information of weapon subtype, `weapsubtype1` gave the numerical categorical info.

- Dropped`nperps`,`weapdetail`,`scite1`,`summary`,`weapsubtype1`,`targsubtype1`,`natlty1`,`weaptype1`,`eventid`,`country`,`region`,`attacktype1`,`targtype1`,`specificity` and`individual` column as they either were not very relevant or had other alternative columns present which imparted same information as them.



## Missing Value Imputation

Next, we looked deeper into the remaining columns for missing value imputation and other modifications:

### Numerical columns

- for numerical columns, imputed missing values of columns `multiple` with 1, `latitude`, `guncertain` and `longitude` with their respective modes (as percentage of missing values was not very high (around 2.5%).
- for columns such as `nperpcap`, `claimed`, `nkill`, `nkillus`, `nkillter`,`nwound`,`nwoundus`,`nwoundte` and `ishostkid`, imputed missing value with **-99** as it is being used for indicating missing value.

### Categorical. columns
- for columns `city`, `targsubtype1_txt`, `corp1`, `target1`, `natlty1_txt` and `weapsubtype1_txt`, imputed missing values with **Missing Info** to avoid any confusion with non-null values.

## Datatype Conversion and Mapping

- After this we changed the datatype of some float columns which were supposed to integer datatype but were misclassfied into the wrong datatype, these columns included `nperpcap`, `nkill`, `nkillus`, `nkillter`, `nwound`, `nwoundus` and `nwoundte`
- For some columns which were having integer values (0, 1 and -99), we mapped them as {1:"Yes", 0:"No", -99:"Unknown", -9:"Unknown"} as per the convention given in `Codebook.pdf`

## Renaming the columns

- Finally, we used the `Codebook.pdf` to rename the columns to more suitable and interpretable names.
- Saved and generated the cleaned data into `gtd_cleaned.csv`